### PR TITLE
Avoid unnecessary allocations in recording rule evaluation

### PR DESCRIPTION
**Pull request by charleskorn**
When evaluating a rule, `RecordingRule.Eval()` uses a `labels.Builder` instance to build the label set for the recorded samples.

`Builder.Labels()` assumes that most of the time, the `Builder` is being used to remove or replace labels in the existing label set rather than add to them. 

But in the case of a recording rule, we're likely adding or replacing labels from the rule definition, so `Builder.Labels()` ends up allocating two `Labels` instances (once [here](https://github.com/prometheus/prometheus/blob/fa0f04bbc6f41708782d4e259fd1fb38896bb904/model/labels/labels.go#L542), and another later on in `append()` [here](https://github.com/prometheus/prometheus/blob/fa0f04bbc6f41708782d4e259fd1fb38896bb904/model/labels/labels.go#L564)). 

Instead, during rule evaluation, we can calculate the upper bound of the length of the label set and allocate just one `Labels` instance ahead of time. This reduces allocations at the risk of slightly overallocating - if the rule overrides some of the labels returned in the query, we won't need quite so much space in the label set.

For one Mimir cluster, we've seen that `Builder.Labels()` is responsible for ~6% of allocations attributable to rule evaluation. (Note that in Mimir, the behaviour is slightly different as we perform a remote query and parse the JSON response, whereas in Prometheus, the query is performed locally.)

There's a similar improvement to be had in `AlertingRule.Eval()` as well - I will address this in a later PR.

## Benchmark results 

Note that this benchmark includes the cost of evaluating the query used in the benchmark's rule.

```
name                                                                                old time/op    new time/op    delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.07µs ± 2%    6.69µs ± 0%  -5.38%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.21µs ± 0%    6.93µs ± 0%  -3.95%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    7.05µs ± 0%    6.70µs ± 1%  -4.89%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               14.3µs ± 1%    14.4µs ± 0%    ~     (p=0.690 n=5+5)

name                                                                                old alloc/op   new alloc/op   delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.13kB ± 0%    6.44kB ± 0%  -9.66%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.13kB ± 0%    6.83kB ± 0%  -4.26%  (p=0.029 n=4+4)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    6.94kB ± 0%    6.44kB ± 0%  -7.16%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               15.2kB ± 0%    14.8kB ± 0%  -2.83%  (p=0.008 n=5+5)

name                                                                                old allocs/op  new allocs/op  delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                     127 ± 0%       119 ± 0%  -6.30%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10               127 ± 0%       121 ± 0%  -4.72%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10       125 ± 0%       119 ± 0%  -4.80%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                  239 ± 0%       233 ± 0%  -2.51%  (p=0.008 n=5+5)
```

### CODE REVIEW

**Comment by charleskorn** [comment]
Fuzzing failed in the CI build, but I can't see any information about what exactly failed - is there some way to get more information?

File: rules/recording.go [hunk 1]
```diff
@@ -89,7 +89,8 @@ func (rule *RecordingRule) Eval(ctx context.Context, ts time.Time, query QueryFu
 			lb.Set(l.Name, l.Value)
 		})
 
-		sample.Metric = lb.Labels(labels.EmptyLabels())
+		labelSetSize := len(sample.Metric) + len(rule.labels) + 1 // additional 1 for rule name
+		sample.Metric = lb.Labels(make(labels.Labels, 0, labelSetSize))
```
Comment (roidelapluie) [inline rules/recording.go:93]: We are working on abstracting labels away. You can not use make assuming it's a slice.

**Comment by roidelapluie** [comment]
Cc @bboreham

File: model/labels/labels.go [hunk 2]
```diff
@@ -348,7 +348,7 @@ func (ls Labels) Map() map[string]string {
 	return m
 }
 
-// EmptyLabels returns n empty Labels value, for convenience.
+// EmptyLabels returns an empty Labels value, for convenience.
```
Comment (bboreham) [inline model/labels/labels.go:351]: Please put unrelated changes in separate PRs.

File: rules/recording.go (hunk 1 above)
Comment (bboreham) [reply rules/recording.go:93]: As @roidelapluie says, this change is conflicting with other work to make `labels.Labels` much smaller, so should not go in as-is.

I imagine moving the `NewBuilder` out of the loop would give some improvement.

Why not just pass `sample.Metric` in to `lb.Labels()`?  That will work for the common case where the rule doesn't add its own labels.

Comment (charleskorn) [reply rules/recording.go]: > I imagine moving the `NewBuilder` out of the loop would give some improvement.
> 
> Why not just pass `sample.Metric` in to `lb.Labels()`? That will work for the common case where the rule doesn't add its own labels.

Good ideas - these both improve things even further:

```
name                                                                                old time/op    new time/op    delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  8.62µs ± 0%    8.16µs ± 0%  -5.26%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            8.76µs ± 0%    8.45µs ± 0%  -3.54%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    8.56µs ± 0%    8.18µs ± 0%  -4.53%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               16.3µs ± 0%    15.9µs ± 0%  -2.08%  (p=0.008 n=5+5)

name                                                                                old alloc/op   new alloc/op   delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.37kB ± 0%    6.68kB ± 0%  -9.35%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.37kB ± 0%    7.07kB ± 0%  -4.12%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    7.18kB ± 0%    6.68kB ± 0%  -6.91%  (p=0.000 n=5+4)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               15.5kB ± 0%    15.1kB ± 0%  -2.79%  (p=0.008 n=5+5)

name                                                                                old allocs/op  new allocs/op  delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                     134 ± 0%       126 ± 0%  -5.97%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10               134 ± 0%       128 ± 0%  -4.48%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10       132 ± 0%       126 ± 0%  -4.55%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                  246 ± 0%       240 ± 0%  -2.44%  (p=0.008 n=5+5)
```

File: rules/recording_test.go [hunk 3]
```diff
@@ -83,6 +83,104 @@ func TestRuleEval(t *testing.T) {
...
+		b.Run(scenario.name, func(b *testing.B) {
+			rule := NewRecordingRule("test_rule", scenario.expr, scenario.ruleLabels)
+
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				result, err := rule.Eval(suite.Context(), ts, EngineQueryFunc(suite.QueryEngine(), suite.Storage()), nil, 0)
+
+				require.NoError(b, err)
+				require.ElementsMatch(b, scenario.expected, result)
```
Comment (bboreham) [inline rules/recording_test.go:178]: Don't do a complex check in the inner-loop of a benchmark; it skews the result.
You could do it before `ResetTimer`, or perhaps better to add these cases to `TestRuleEval`.
I think the cases can be shared between Test and Benchmark.

**Review by bboreham — COMMENTED** [review]
Code looks great now, but I have a comment on the benchmark.

File: rules/recording_test.go (hunk 3 above)
Comment (charleskorn) [reply rules/recording_test.go]: Good point, I've made this change in 054dfc3.

Comment (charleskorn) [reply rules/recording_test.go]: The benchmark results now look like this, having re-benchmarked the original and new code with the modified benchmark:

```
name                                                                                old time/op    new time/op    delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.07µs ± 2%    6.69µs ± 0%  -5.38%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.21µs ± 0%    6.93µs ± 0%  -3.95%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    7.05µs ± 0%    6.70µs ± 1%  -4.89%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               14.3µs ± 1%    14.4µs ± 0%    ~     (p=0.690 n=5+5)

name                                                                                old alloc/op   new alloc/op   delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.13kB ± 0%    6.44kB ± 0%  -9.66%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.13kB ± 0%    6.83kB ± 0%  -4.26%  (p=0.029 n=4+4)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    6.94kB ± 0%    6.44kB ± 0%  -7.16%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               15.2kB ± 0%    14.8kB ± 0%  -2.83%  (p=0.008 n=5+5)

name                                                                                old allocs/op  new allocs/op  delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                     127 ± 0%       119 ± 0%  -6.30%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10               127 ± 0%       121 ± 0%  -4.72%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10       125 ± 0%       119 ± 0%  -4.80%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                  239 ± 0%       233 ± 0%  -2.51%  (p=0.008 n=5+5)
```

File: rules/recording_test.go [hunk 4]
```diff
@@ -27,59 +27,128 @@ import (
 	"github.com/prometheus/prometheus/util/teststorage"
 )
 
+type testScenario struct {
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply rules/recording_test.go]: Resolved in b655bfdd3.

File: rules/recording_test.go [hunk 5]
```diff
@@ -27,59 +27,128 @@ import (
...
+	expected   promql.Vector
+}
+
+var (
+	ruleEvaluationTime       = time.Unix(0, 0).UTC()
+	exprWithMetricName, _    = parser.ParseExpr(`metric`)
+	exprWithoutMetricName, _ = parser.ParseExpr(`metric + metric`)
+)
+
+var scenarios = []testScenario{
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply rules/recording_test.go]: Resolved in https://github.com/prometheus/prometheus/commit/b655bfdd3ea82cbee52c2a01fda1b4ec91f79427.

**Review by bboreham — APPROVED** [review]
LGTM

**Comment by charleskorn** [comment]
Are any further changes required? Would be great to get this merged.

**Comment by roidelapluie** [comment]
I would like to first merge the label string behind a build tag

**Comment by roidelapluie** [comment]
Can you rebase this?

**Comment by charleskorn** [comment]
> Can you rebase this?

Yep, done.

**Comment by charleskorn** [comment]
Looks like the Windows build is failing due to non-deterministic order of series in query results - is this expected?

**Comment by bboreham** [comment]
> Looks like the Windows build is failing due to non-deterministic order of series in query results - is this expected?

Prometheus does not guarantee the order of results of an instant query, so yes.
Adding `sort` to the query is one option, or using more sophisticated code to check the result.

**Review by bboreham — COMMENTED** [review]
lgtm, thanks!

**Review by bboreham — APPROVED** [review]
